src/include/serial_inline.h

```diff
@@ -300,32 +299,16 @@ __wt_update_serial(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt, WT_PAGE *page
         return (0);
 
     /*
-     * We would like to call __wt_txn_update_oldest only in the event that there are further updates
-     * to this page, the check against WT_TXN_NONE is used as an indicator of there being further
-     * updates on this page.
+     * Look for obsolete updates if:
+     * 1. The page has been configured for eviction soon. It is possible for a hot page
+     *    that is marked for evicting soon but cannot be able to evicted due to the
+     *    concurrent access. Removing the obsolete updates from these pages can reduce
+     *    the page memory footprint and avoids the unnecessary forced eviction.
+     * 2. There are several updates in the cache that need to be removed.
      */
-    if ((txn = page->modify->obsolete_check_txn) != WT_TXN_NONE) {
-        obsolete_timestamp = page->modify->obsolete_check_timestamp;
-        if (!__wt_txn_visible_all(session, txn, obsolete_timestamp)) {
-            /* Try to move the oldest ID forward and re-check. */
-            ret = __wt_txn_update_oldest(session, 0);
-            /*
-             * We cannot proceed if we fail here as we have inserted the updates to the update
-             * chain. Panic instead. Currently, we don't ever return any error from
-             * __wt_txn_visible_all. We can catch it if we start to do so in the future and properly
-             * handle it.
-             */
-            if (ret != 0)
-                WT_RET_PANIC(session, ret, "fail to update oldest after serializing the updates");
-
-            if (!__wt_txn_visible_all(session, txn, obsolete_timestamp))
-                return (0);
-        }
-
-        page->modify->obsolete_check_txn = WT_TXN_NONE;
-    }
-
-    __wt_update_obsolete_check(session, cbt, upd->next, true);
+    if (WT_READGEN_EVICT_SOON(page->read_gen) ||
```

Comment:
Instead of this check, can we do something that compare the size of total updates to the page size? If the page size is larger than half of the force eviction threshold and the size of updates is larger than half of the page size, we  trim the updates?

Reply:
I want to avoid calling the obsolete check from normal update operations unless it is required. If we add the size check, whenever the page crosses half of the page size, we look for obsolete updates, which can cause performance problems.

If we randomly pick the obsolete updates after a couple of updates, we can still miss the obsolete updates of the keys and look for the wrong key.

In this way, whenever the page is marked for urgent eviction, we look for obsolete updates for every key that is modified. If the reconciliation happens regularly on these pages, we do not need to clear in this flow.

Reply:
It is difficult for me to link WT_READGEN_EVICT_SOON to forced eviction. But I understand now. Thanks.